### PR TITLE
Allow interacting with submissions

### LINKED ISSUE #54: Interactive Mode
Perhaps we should make the toolkit drop to a shell on programs which require interaction? Or, in the case of a failed run, perhaps drop to a shell for inspection?

Of course, we can't really do this with parallel evaluation, unless we go through all of the submissions and then re-run the ones that failed in interactive mode.

I'd be open to suggestions on this.

### PR DESCRIPTION
**Pull request by hawkrives**
TL;DR: `cs251tk --interact`.

---

Some submissions require interaction with the program to properly evaluate them, such as the story assignments from homeworks 4 and 5.

This PR adds the `--interact` flag, which runs one submission at a time, dropping you into a psuedo-tty (I think) where you can interact with the submitted program.

When the program is done, you can send EOF (<kbd>^D</kbd>, control-D) to exit the interaction.

After each interaction, you can easily run the submission again by answering the question "Do you want to run the submission again? [y/N]", which defaults to "no". (Examples of when you might want to run it again: `countup.cpp`, `countdown.cpp`.)

This mode offers no infinite loop protection, but <kbd>^C</kbd> should exit the submitted program while keeping the toolkit alive. In theory.

This mode will be refined based on feedback from the graders.

Closes #54.

### CODE REVIEW

**Comment by hawkrives** [comment]
Example log file:

-----

# hw4 – example

## iotest.cpp (Tue Feb 16 17:38:21 2016 -0600)

    // Inputs, Outputs, String Streams practice
    #include <iostream>
    #include <string>
    #include <sstream>
    using namespace std;

    int main ()
    {
      string name;     //user's name
      string mystr;    //string identifier
      float shoesize;  //user's shoe size
      float number;    //user's favorite number

      cout << "What is your name?: ";
      getline (cin,name);
      cout << "What is your shoe size?: ";
      getline (cin,mystr);
      stringstream(mystr) >> shoesize;
      cout << "What is your favorite number?: ";
      getline (cin,mystr);
      stringstream(mystr) >> number;
      cout << name << ", your result is " << (shoesize + number) / 2 << endl;

      return 0;
    }


**no warnings: `g++ --std=c++11 ./iotest.cpp -o ./iotest.cpp.exec`**


**results of `['./iotest.cpp.exec']`** (status: success)

    What is your name?: Hawken
    What is your shoe size?: 11
    What is your favorite number?: 11
    Hawken, your result is 11
**results of `['./iotest.cpp.exec']`** (status: success)

    What is your name?: Hawken
    What is your shoe size?: 15
    What is your favorite number?: 0
    Hawken, your result is 7.5
**results of `['./iotest.cpp.exec']`** (status: success)

    What is your name?: 000
    What is your shoe size?: null
    What is your favorite number?: null
    000, your result is 0



## countdown.cpp (Tue Feb 16 17:38:21 2016 -0600)

    // Countdown to show While Loop

    #include <iostream>
    using namespace std;

    int main ()
    {
      int n;
      cout << "Enter the starting number > ";
      cin >> n;

      while (n>-11) /*modified to make the countdown end with -10, FIRE! */  {
        cout << n << ", ";
        --n;
      }

      cout << "FIRE!\n";
      return 0;
    }


**no warnings: `g++ --std=c++11 ./countdown.cpp -o ./countdown.cpp.exec`**


**results of `['./countdown.cpp.exec']`** (status: success)

    Enter the starting number > 15
    15, 14, 13, 12, 11, 10, 9, 8, 7, 6, 5, 4, 3, 2, 1, 0, -1, -2, -3, -4, -5, -6, -7, -8, -9, -10, FIRE!
**results of `['./countdown.cpp.exec']`** (status: success)

    Enter the starting number > -11
    FIRE!
**results of `['./countdown.cpp.exec']`** (status: success)

    Enter the starting number > 0 -9
    -9, -10, FIRE!



## countup.cpp (Tue Feb 16 17:38:21 2016 -0600)

    //Count Up to practice While Loops

    #include <iostream>
    using namespace std;

    int main ()
    {
      int n;
      cout << "Enter the starting number > ";
      cin >> n;

      while (n<21) {
        cout << n << ", ";
        ++n;
      }

      cout << "Done!" << endl;
      return 0;
    }


**no warnings: `g++ --std=c++11 ./countup.cpp -o ./countup.cpp.exec`**


**results of `['./countup.cpp.exec']`** (status: success)

    Enter the starting number > 15
    15, 16, 17, 18, 19, 20, Done!



## story.cpp (Tue Feb 16 17:38:21 2016 -0600)

    //Writing a story

    #include <iostream>
    #include <string>
    #include <sstream>
    using namespace std;

    int main ()
    {
      /* Introduction:*/
      cout << "Once upon a time, there was a mouse named Fred who loved to roam St. Olaf College campus... \n";

      /*Declare Variables for choices*/
      string mystr;
      int choice1;
      int choice2a;
      int choice2b;

      /*Choice 1*/
      cout << "Please select one of the following options:\n" << "1) go to Dittman Art  Center\n" << "2) go to Regents Hall of Natural Sciences\n";
      cout << "Your response: ";
      getline (cin,mystr);
      stringstream(mystr) >> choice1;

      if (choice1==1) /*Choice 2: Dittman*/ {
        cout << "Fred knew all the Art students were always looking for inspiration for their work, and decided to help...\n";
        cout << "Please select one of the following options:\n" << "1) find a paint bucket\n" << "2) make some music\n";
        cout << "Your response: ";
        getline (cin,mystr);
        stringstream(mystr) >> choice2a;

        /*Endings*/
        if (choice2a==1) {
          cout << "Fred tipped over a paint bucket of his favorite color. The paint got all over a student's painting, but luckily the professor thought it added a really interesting modern element! Good job, Fred!" << endl;
        } else {
          cout << "Fred wanted to give the dancers some mouse-like inspiration, so he decided to make some percussion in the air vents for them to dance to! Unfortunately it seemed to scare them. The End." << endl;
        }
      } else /*Choice 2: Regents*/ {
        cout << "Fred loves the science labs. After visiting his captive buddies in the Psych department, he usually finds his way to...";
        cout << "Please choose one of the following options:\n" << "1) go to the chemistry lab\n" << "2) go to RNS 390\n";
        cout << "Your response: ";
        getline (cin,mystr);
        stringstream(mystr) >> choice2b;

        /*Endings*/
        if (choice2b==1) {
          cout << "Fred, like an other mouse, appreciated a drink at the end of long days. The chem labs had plenty of chemicals to help him forget his sorrows, but this time he chose the wrong chemical to try. The chemistry students held a short and sweet funeral for Fred the next morning. The End." << endl;
        } else {
          cout << "Fred loved hanging out with students! What better way to do so than to go to a study session and make some new friends? The biology students were friendly and kept running around with Fred ina game of tag! Fred must have been 'it' though, because all the students kept avoiding let him touch them. The End." << endl;
        }
      }
    }


**no warnings: `g++ --std=c++11 ./story.cpp -o ./story.cpp.exec`**


**results of `['./story.cpp.exec']`** (status: success)

    Once upon a time, there was a mouse named Fred who loved to roam St. Olaf College campus... 
    Please select one of the following options:
    1) go to Dittman Art  Center
    2) go to Regents Hall of Natural Sciences
    Your response: 1
    Fred knew all the Art students were always looking for inspiration for their work, and decided to help...
    Please select one of the following options:
    1) find a paint bucket
    2) make some music
    Your response: 1
    Fred tipped over a paint bucket of his favorite color. The paint got all over a student's painting, but luckily the professor thought it added a really interesting modern element! Good job, Fred!
**results of `['./story.cpp.exec']`** (status: success)

    Once upon a time, there was a mouse named Fred who loved to roam St. Olaf College campus... 
    Please select one of the following options:
    1) go to Dittman Art  Center
    2) go to Regents Hall of Natural Sciences
    Your response: 1
    Fred knew all the Art students were always looking for inspiration for their work, and decided to help...
    Please select one of the following options:
    1) find a paint bucket
    2) make some music
    Your response: 2
    Fred wanted to give the dancers some mouse-like inspiration, so he decided to make some percussion in the air vents for them to dance to! Unfortunately it seemed to scare them. The End.

**Comment by hawkrives** [comment]
I have updated the PR to print the file that is being run.

```
Recording ['./iotest.cpp.exec']. Send EOF (^D) to end.
```

**Comment by hawkrives** [comment]
There's probably something that I should test here, but I don't care enough right now…

**Comment by rye** [comment]
One thought that I had was to be able to drop to a shell if something fails—for inspection, that is. Not sure how safe or useful that could be, but it would definitely be nice in my opinion. Either way, this PR looks good to me.

**Comment by hawkrives** [comment]
Yeah, that'd be nice-to-have, definitely. 

It'd be a good deal trickier, since most of the time when things go wrong you'll be in parallel mode, and we can't exit that to switch to interactive mode once it's begun.

What about adding a switch to prevent the toolkit from restoring the initial state when it finishes an assignment? So you can go look at the folders after the toolkit has finished a run?

**Comment by hawkrives** [comment]
Which, rereading https://github.com/StoDevX/cs251-toolkit/issues/3, it sounds like I think that the toolkit already doesn't reset the folders after it runs; only beforehand. So you should be able to just use the normal shell.
